pkg/kapis/openpitrix/v1/handler.go

```diff
@@ -1,76 +1,967 @@
+		return
+	}
+
+	err = h.openpitrix.ModifyCategory(categoryId, &modifyCategoryRequest)
+
+	if status.Code(err) == codes.InvalidArgument {
+		resp.WriteHeaderAndEntity(http.StatusBadRequest, errors.Wrap(err))
+		return
+	}
+	if err != nil {
+		resp.WriteHeaderAndEntity(http.StatusInternalServerError, errors.Wrap(err))
+		return
+	}
+
+	resp.WriteEntity(errors.None)
+}
+func (h *openpitrixHandler) DescribeCategory(req *restful.Request, resp *restful.Response) {
+	categoryId := req.PathParameter("category")
+
+	result, err := h.openpitrix.DescribeCategory(categoryId)
+
+	if status.Code(err) == codes.NotFound {
+		resp.WriteHeaderAndEntity(http.StatusNotFound, errors.Wrap(err))
+		return
+	}
+
+	if err != nil {
+		resp.WriteHeaderAndEntity(http.StatusInternalServerError, errors.Wrap(err))
+		return
+	}
+
+	resp.WriteEntity(result)
+}
+func (h *openpitrixHandler) ListCategories(req *restful.Request, resp *restful.Response) {
+	limit, offset := params.ParsePaging(req)
+	orderBy := params.GetStringValueWithDefault(req, params.OrderByParam, openpitrix.CreateTime)
+	reverse := params.GetBoolValueWithDefault(req, params.ReverseParam, true)
+	statistics := params.GetBoolValueWithDefault(req, "statistics", false)
+	conditions, err := params.ParseConditions(req)
+
+	if err != nil {
+		resp.WriteHeaderAndEntity(http.StatusBadRequest, errors.Wrap(err))
+		return
+	}
+
+	result, err := h.openpitrix.ListCategories(conditions, orderBy, reverse, limit, offset)
+
+	if err != nil {
+		klog.Errorln(err)
+		resp.WriteHeaderAndEntity(http.StatusInternalServerError, errors.Wrap(err))
+		return
+	}
+
+	if statistics {
+		for _, item := range result.Items {
+			if category, ok := item.(*openpitrix.Category); ok {
+				statisticsResult, err := h.openpitrix.ListApps(&params.Conditions{Match: map[string]string{"category_id": category.CategoryID, "status": openpitrix.StatusActive, "repo": openpitrix.BuiltinRepoId}}, "", false, 0, 0)
+				if err != nil {
+					klog.Errorln(err)
+					resp.WriteHeaderAndEntity(http.StatusInternalServerError, errors.Wrap(err))
+					return
+				}
+				category.AppTotal = &statisticsResult.TotalCount
+			}
+		}
+	}
+
+	resp.WriteEntity(result)
+}
+
+func (h *openpitrixHandler) CreateRepo(req *restful.Request, resp *restful.Response) {
+	createRepoRequest := &openpitrix.CreateRepoRequest{}
+	err := req.ReadEntity(createRepoRequest)
+	if err != nil {
+		resp.WriteHeaderAndEntity(http.StatusBadRequest, errors.Wrap(err))
+		return
+	}
+	validate, _ := strconv.ParseBool(req.QueryParameter("validate"))
+
+	var result interface{}
+
+	if validate {
+		validateRepoRequest := &openpitrix.ValidateRepoRequest{
+			Type:       createRepoRequest.Type,
+			Url:        createRepoRequest.URL,
+			Credential: createRepoRequest.Credential,
+		}
+		result, err = h.openpitrix.ValidateRepo(validateRepoRequest)
+	} else {
+		result, err = h.openpitrix.CreateRepo(createRepoRequest)
+	}
+
+	if status.Code(err) == codes.InvalidArgument {
```

Comment:
what will happen to status.Code(err) if err == nil ? I think it's better to be rewritten like following:
```
if err != nil {
  if status.Code(err) == codes.InvalidArgument {
     ...
     return
  }
  api.HandleInternalError(resp, err)
  return
}
```